hugr-core/src/import.rs

```diff
@@ -0,0 +1,1214 @@
+    fn import_type_arg(&mut self, term_id: model::TermId) -> Result<TypeArg, ImportError> {
+        match self.get_term(term_id)? {
+            model::Term::Wildcard => Err(error_uninferred!("wildcard")),
+            model::Term::Apply { .. } => {
+                Err(error_uninferred!("application with implicit parameters"))
+            }
+
+            model::Term::Var(var) => {
+                let (index, var_type) = self.resolve_local_ref(var)?;
+                let decl = self.import_type_param(var_type)?;
+                Ok(TypeArg::new_var_use(index, decl))
+            }
+
+            model::Term::List { .. } => {
+                let elems = self
+                    .import_closed_list(term_id)?
+                    .iter()
+                    .map(|item| self.import_type_arg(*item))
+                    .collect::<Result<_, _>>()?;
+
+                Ok(TypeArg::Sequence { elems })
+            }
+
+            model::Term::Str(value) => Ok(TypeArg::String {
+                arg: value.to_string(),
+            }),
+
+            model::Term::Quote { .. } => Ok(TypeArg::Type {
+                ty: self.import_type(term_id)?,
+            }),
+            model::Term::Nat(value) => Ok(TypeArg::BoundedNat { n: *value }),
+            model::Term::ExtSet { .. } => Ok(TypeArg::Extensions {
+                es: self.import_extension_set(term_id)?,
+            }),
+
+            model::Term::StrType => Err(error_unsupported!("`str` as `TypeArg`")),
+            model::Term::NatType => Err(error_unsupported!("`nat` as `TypeArg`")),
+            model::Term::ListType { .. } => Err(error_unsupported!("`(list ...)` as `TypeArg`")),
+            model::Term::ExtSetType => Err(error_unsupported!("`ext-set` as `TypeArg`")),
+            model::Term::Type => Err(error_unsupported!("`type` as `TypeArg`")),
+            model::Term::ApplyFull { .. } => Err(error_unsupported!("custom types as `TypeArg`")),
+            model::Term::Constraint => Err(error_unsupported!("`constraint` as `TypeArg`")),
+            model::Term::StaticType => Err(error_unsupported!("`static` as `TypeArg`")),
+            model::Term::ControlType => Err(error_unsupported!("`ctrl` as `TypeArg`")),
+
+            model::Term::FuncType { .. }
+            | model::Term::Adt { .. }
+            | model::Term::Control { .. } => Err(model::ModelError::TypeError(term_id).into()),
+        }
+    }
+
+    fn import_extension_set(
+        &mut self,
+        term_id: model::TermId,
+    ) -> Result<ExtensionSet, ImportError> {
+        match self.get_term(term_id)? {
+            model::Term::Wildcard => Err(error_uninferred!("wildcard")),
+
+            model::Term::Var(var) => {
+                let mut es = ExtensionSet::new();
+                let (index, _) = self.resolve_local_ref(var)?;
+                es.insert_type_var(index);
+                Ok(es)
+            }
+
+            model::Term::ExtSet { extensions, rest } => {
```

Comment:
It's for extension sets where not all extensions are known, so the rest can be a variable. It's like the tail of a list.

There's a mismatch here with the core: The core at the moment allows variables in rows (becoming lists in the model since that's the name of these things in type theory. That also opens up using "row" to refer to actual rows if we want to add them) and in extension sets. But having multiple variables in there leads to a weird type theory in which unification is ambiguous. When we want to union multiple unknown extension sets or lists, we can do that with constraints.